extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java

```diff
@@ -210,4 +211,18 @@ public void handle(RoutingContext event) {
             }
         }
     }
+
+    private static boolean isSameOrigin(HttpServerRequest request, String origin) {
+        String absUriString = request.absoluteURI();
+        if (absUriString.startsWith(origin)) {
+            // Make sure that Origin URI contains scheme, host, and port.
+            // If no port is set in Origin URI then the request URI must not have it set either
+            URI baseUri = URI.create(absUriString.substring(0, origin.length()));
```

Comment:
IIRC, parsing URLs has a cost that is not exactly neutral.

Reply:
@gsmet I agree, but I'm not sure how to handle it otherwise, I was not sure about `request.host()`, will it return what is in the URI or in the HOST header - in the latter case it is not guaranteed to natch `request.scheme()` and `request.port()` probably parse it as well.
FYI, the cost can be avoided if the local Origin is configured so the users have options

Reply:
getting the actual parts sounds reasonable - how does vert.x cors handle this?

Reply:
Right, just a `startsWith` check alone would prove nothing which will accept incomplete Origin values

Reply:
Yes, so my question is how does vert.x cors handle this ? I assume they need similar split of the elements.

Reply:
@cescoffier,  @pmlopes Does Vert.x do any special treatment for the same Origin requests ? 

I think, even though it is true that converting a String to URI is not necessarily negligible cost-wise,  this will only be the minor issue for the same Origin POST requests, simple/same origin GET are not having Origin added by the browser, never, AFAIK, so the impact on the typical same host operations would be very minimal. 

So IMHO it is OK to have this parsing code in place - unfortunately we can't now just let the request go if the Origin has proved to be unmatched/unverified. 

And as I said, users will have an option to list the same Origin in the CORS filter configuration.
The only case where it can indeed become a bit annoying is when the setup is quite complex - which combined with the fact that it only affects POST/PUT is probably makes it a 10% case.

If we could this checks more optimal though then it would indeed be nice, so if Paulo or Clement can recommend something it would be great. Also CC @stuartwdouglas

Reply:
@sberyozkin i went and looked in vert.x and found https://github.com/vert-x3/vertx-web/blob/master/vertx-web/src/main/java/io/vertx/ext/web/impl/Origin.java#L148

Its a custom parser.